src/pathfinding.cpp

```diff
@@ -0,0 +1,352 @@
+            layer.state[index] == ASL_CLOSED ) {
+            return;
+        }
+
+        layer.state [index] = ASL_OPEN;
+        layer.gscore[index] = gscore;
+        layer.parent[index] = from;
+        layer.score [index] = score;
+        open.push( std::make_pair( score, to ) );
+    }
+
+    void close_point( const tripoint &p ) {
+        auto &layer = get_layer( p.z );
+        const int index = flat_index( p.x, p.y );
+        layer.state[index] = ASL_CLOSED;
+    }
+};
+
+// Returns a tile with `flag` in the overmap tile that `t` is on
+template<ter_bitflags flag>
+tripoint vertical_move_destination( const map &m, const tripoint &t )
+{
+    if( !m.has_zlevels() ) {
+        return tripoint_min;
+    }
+
+    constexpr int omtileszx = SEEX * 2;
+    constexpr int omtileszy = SEEY * 2;
+    real_coords rc( m.getabs( t.x, t.y ) );
+    point omtile_align_start(
+        m.getlocal( rc.begin_om_pos() )
+    );
+
+    tripoint from( omtile_align_start.x, omtile_align_start.y, t.z );
+    tripoint to( omtile_align_start.x + omtileszx, omtile_align_start.y + omtileszy, t.z );
+
+    // TODO: Avoid up to 576 bounds checks by using methods that don't check bounds
+    for( const tripoint &p : m.points_in_rectangle( from, to ) ) {
+        if( m.has_flag( flag, p ) ) {
+            return p;
+        }
+    }
+
+    return tripoint_min;
+}
+
+std::vector<tripoint> map::route( const tripoint &f, const tripoint &t,
+                                  const int bash, const int maxdist ) const
+{
+    /* TODO: If the origin or destination is out of bound, figure out the closest
+     * in-bounds point and go to that, then to the real origin/destination.
+     */
+
+    int linet1 = 0, linet2 = 0;
+    if( !inbounds( f ) || !inbounds( t ) ) {
+        // Note: The creature needs to understand not-moving upwards
+        // or else the plans can cause it to do so.
+        if( sees( f, t, -1, linet1, linet2 ) ) {
+            return line_to( f, t, linet1, linet2 );
+        } else {
+            std::vector<tripoint> empty;
+            return empty;
+        }
+    }
+    // First, check for a simple straight line on flat ground
+    // Except when the player is on the line - we need to do regular pathing then
+    const tripoint &pl_pos = g->u.pos();
+    if( f.z == t.z && clear_path( f, t, -1, 2, 2, linet1, linet2 ) ) {
+        const auto line_path = line_to( f, t, linet1, linet2 );
+        if( pl_pos.z != f.z ) {
+            // Player on different z-level, certainly not on the line
+            return line_path;
+        }
+
+        bool is_ok = true;
+        for( auto &line_pt : line_path ) {
+            if( line_pt == pl_pos ) {
+                is_ok = false;
+                break;
+            }
+        }
+        if( is_ok ) {
+            return line_path;
+        }
+    }
+
+    const int pad = 8;  // Should be much bigger - low value makes pathfinders dumb!
+    int minx = f.x - pad;
```

Comment:
The whole logic for the min* variables is to cumbersome, just get the minimum and subtract pad:
```C++
int minx = std::min( f.x, t.x ) - pad;
```